File: src/user_examples/cs_user_boundary_conditions-auto_inlet_profile.f90
!-------------------------------------------------------------------------------

!VERS

! This file is part of code_saturne, a general-purpose CFD tool.
!
! Copyright (C) 1998-2022 EDF S.A.
!
! This program is free software; you can redistribute it and/or modify it under
! the terms of the GNU General Public License as published by the Free Software
! Foundation; either version 2 of the License, or (at your option) any later
! version.
!
! This program is distributed in the hope that it will be useful, but WITHOUT
! ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
! FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
! details.
!
! You should have received a copy of the GNU General Public License along with
! this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
! Street, Fifth Floor, Boston, MA 02110-1301, USA.

!-------------------------------------------------------------------------------

!===============================================================================
! Function:
! ---------

! Example of cs_f_user_boundary_conditions subroutine.f90 for inlet
! with automatic inlet profile.

! This example assumes the mesh is orthogonal at the inlet.
!
!-------------------------------------------------------------------------------

!-------------------------------------------------------------------------------
! Arguments
!______________________________________________________________________________.
!  mode           name          role                                           !
!______________________________________________________________________________!
!> \param[in]     nvar          total number of variables
!> \param[in]     nscal         total number of scalars
!> \param[out]    icodcl        boundary condition code:
!>                               - 1 Dirichlet
!>                               - 2 Radiative outlet
!>                               - 3 Neumann
!>                               - 4 sliding and
!>                                 \f$ \vect{u} \cdot \vect{n} = 0 \f$
!>                               - 5 smooth wall and
!>                                 \f$ \vect{u} \cdot \vect{n} = 0 \f$
!>                               - 6 rough wall and
!>                                 \f$ \vect{u} \cdot \vect{n} = 0 \f$
!>                               - 9 free inlet/outlet
!>                                 (input mass flux blocked to 0)
!>                               - 13 Dirichlet for the advection operator and
!>                                    Neumann for the diffusion operator
!> \param[in]     itrifb        indirection for boundary faces ordering
!> \param[in,out] itypfb        boundary face types
!> \param[in,out] izfppp        boundary face zone number
!> \param[in]     dt            time step (per cell)
!> \param[in,out] rcodcl        boundary condition values:
!>                               - rcodcl(1) value of the dirichlet
!>                               - rcodcl(2) value of the exterior exchange
!>                                 coefficient (infinite if no exchange)
!>                               - rcodcl(3) value flux density
!>                                 (negative if gain) in w/m2
!>                                 -# for the velocity \f$ (\mu+\mu_T)
!>                                    \gradt \, \vect{u} \cdot \vect{n}  \f$
!>                                 -# for the pressure \f$ \Delta t
!>                                    \grad P \cdot \vect{n}  \f$
!>                                 -# for a scalar \f$ cp \left( K +
!>                                     \dfrac{K_T}{\sigma_T} \right)
!>                                     \grad T \cdot \vect{n} \f$
!_______________________________________________________________________________

subroutine cs_f_user_boundary_conditions &
 ( nvar   , nscal  ,                                              &
   icodcl , itrifb , itypfb , izfppp ,                            &
   dt     ,                                                       &
   rcodcl )

!===============================================================================

!===============================================================================
! Module files
!===============================================================================

use paramx
use numvar
use optcal
use cstphy
use cstnum
use entsor
use parall
use period
use ppppar
use ppthch
use coincl
use cpincl
use ppincl
use ppcpfu
use atincl
use atsoil
use ctincl
use cs_fuel_incl
use mesh
use field
use turbomachinery
use cs_c_bindings

!===============================================================================

implicit none

! Arguments

integer          nvar   , nscal

integer          icodcl(nfabor,nvar)
integer          itrifb(nfabor), itypfb(nfabor)
integer          izfppp(nfabor)

double precision dt(ncelet)
double precision rcodcl(nfabor,nvar,3)

! Local variables

!< [loc_var_dec]
integer          ifac, iel, ii, ilelt, nlelt

double precision xustar2, xdh, d2s3, rhomoy
double precision acc(2), fmprsc, fmul, uref2, vnrm

integer, allocatable, dimension(:) :: lstelt, mrkcel

double precision, dimension(:), pointer :: bfpro_rom
double precision, dimension(:), pointer :: cvar_k, cvar_ep, cvar_phi
double precision, dimension(:), pointer :: cvar_omg
double precision, dimension(:), pointer :: cvar_al, cvar_fb
double precision, dimension(:), pointer :: cvar_nusa
double precision, dimension(:), pointer :: cvar_scal
double precision, dimension(:,:), pointer :: cvar_vel
double precision, dimension(:,:), pointer :: cvar_rij
!< [loc_var_dec]

!===============================================================================
! Initialization
!===============================================================================

!< [init]
allocate(lstelt(nfabor))  ! temporary array for boundary faces selection

! Map field arrays
call field_get_val_v(ivarfl(iu), cvar_vel)
call field_get_val_s(ibrom, bfpro_rom)

if (itytur.eq.2) then

  call field_get_val_s(ivarfl(ik), cvar_k)
  call field_get_val_s(ivarfl(iep), cvar_ep)

elseif (itytur.eq.3) then
  call field_get_val_v(ivarfl(irij), cvar_rij)
  call field_get_val_s(ivarfl(iep), cvar_ep)

  if (iturb.eq.32) then
    call field_get_val_s(ivarfl(ial), cvar_al)
  endif

elseif (itytur.eq.5) then

  call field_get_val_s(ivarfl(ik), cvar_k)
  call field_get_val_s(ivarfl(iep), cvar_ep)
  call field_get_val_s(ivarfl(iphi), cvar_phi)

  if (iturb.eq.50) then
    call field_get_val_s(ivarfl(ifb), cvar_fb)
  elseif (iturb.eq.51) then
    call field_get_val_s(ivarfl(ial), cvar_al)
  endif

elseif (iturb.eq.60) then

  call field_get_val_s(ivarfl(ik), cvar_k)
  call field_get_val_s(ivarfl(iomg), cvar_omg)

elseif (iturb.eq.70) then

  call field_get_val_s(ivarfl(inusa), cvar_nusa)

endif

d2s3 = 2.d0/3.d0
!< [init]

!===============================================================================
! Assign a pseudo-periodic channel type inlet to a set of boundary faces.

! For each subset:
! - use selection criteria to filter boundary faces of a given subset
! - loop on faces from a subset
!   - set the boundary condition for each face

! A feedback loop is used so as to progressively reach a state similar
! to that of a periodic channel at the inlet.
!===============================================================================

!< [example_1]
call getfbr('INLET', nlelt, lstelt)
!==========

fmprsc = 1.d0 ! mean prescribed velocity

if (ntcabs.eq.1) then

  ! For the Rij-EBRSM model (and possibly V2f), we need a non-flat profile,
  ! so as to ensure turbulent production, and avoid laminarization;
  ! here, we simply divide the initial velocity by 10 for inlet
  ! faces adjacent to the wall.

  ! The loop below assumes wall conditions have been defined first
  ! (in the GUI, or in this file, before the current test).

  if (iturb.eq.32 .or. itytur.eq.5) then

    allocate(mrkcel(ncelet))
    do iel = 1, ncelet
      mrkcel(iel) = 0
    enddo

    do ifac = 1, nfabor
      if (itypfb(ifac) .eq. iparoi) then
        iel = ifabor(ifac)
        mrkcel(iel) = 1
      endif
    enddo

  endif

  do ilelt = 1, nlelt

    ifac = lstelt(ilelt)
    iel = ifabor(ifac)

    itypfb(ifac) = ientre

    rcodcl(ifac,iu,1) = - fmprsc * surfbo(1,ifac) / surfbn(ifac)
    rcodcl(ifac,iv,1) = - fmprsc * surfbo(2,ifac) / surfbn(ifac)
    rcodcl(ifac,iw,1) = - fmprsc * surfbo(3,ifac) / surfbn(ifac)

    if (iturb.eq.32 .or. itytur.eq.5) then
      if (mrkcel(iel) .eq. 1) then
        rcodcl(ifac,iu,1) = fmprsc/10.d0
      endif
    endif

    uref2 = rcodcl(ifac,iu,1)**2  &
          + rcodcl(ifac,iv,1)**2  &
          + rcodcl(ifac,iw,1)**2
    uref2 = max(uref2,1.d-12)

    !   Turbulence example computed using equations valid for a pipe.

    !   We will be careful to specify a hydraulic diameter adapted
    !     to the current inlet.

    !   We will also be careful if necessary to use a more precise
    !     formula for the dynamic viscosity use in the calculation of
    !     the Reynolds number (especially if it is variable, it may be
    !     useful to take the law from 'usphyv'. Here, we use by default
    !     the 'viscl0" value.
    !   Regarding the density, we have access to its value at boundary
    !     faces (romb) so this value is the one used here (specifically,
    !     it is consistent with the processing in 'usphyv', in case of
    !     variable density)

    !     Hydraulic diameter
    xdh     = 1.d0

    ! Calculation of turbulent inlet conditions using
    !   the turbulence intensity and standard laws for a circular pipe
    !   (their initialization is not needed here but is good practice)

    rhomoy  = bfpro_rom(ifac)

    call turbulence_bc_inlet_hyd_diam(ifac, uref2, xdh, rhomoy, viscl0,  &
                                      rcodcl)

    ! Handle scalars
    if (nscal.gt.0) then
      do ii = 1, nscal
        rcodcl(ifac,isca(ii),1) = 1.d0
      enddo
    endif

  enddo

  if (iturb.eq.32 .or. itytur.eq.5) then
    deallocate(mrkcel)
  endif

else

! Subsequent time steps
!----------------------

  acc(1) = 0.d0
  acc(2) = 0.d0

  ! Estimate multiplier

  do ilelt = 1, nlelt

    ifac = lstelt(ilelt)
    iel = ifabor(ifac)

    vnrm = sqrt(cvar_vel(1,iel)**2 + cvar_vel(2,iel)**2 + cvar_vel(3,iel)**2)
    acc(1) = acc(1) + vnrm*surfbn(ifac)
    acc(2) = acc(2) + surfbn(ifac)

  enddo

  if (irangp.ge.0) then
    call parrsm(2, acc)
  endif

  if (acc(1).le.epzero) then
     fmul = 0.d0 ! zero velocity in bulk domain
  else
     fmul = fmprsc/(acc(1)/acc(2)) ! 1 / estimate flow multiplier
  endif

  ! Apply BC

  do ilelt = 1, nlelt

    ifac = lstelt(ilelt)
    iel = ifabor(ifac)

    itypfb(ifac) = ientre

    vnrm = sqrt(cvar_vel(1,iel)**2 + cvar_vel(2,iel)**2 + cvar_vel(3,iel)**2)

    rcodcl(ifac,iu,1) = - fmul * vnrm * surfbo(1,ifac) / surfbn(ifac)
    rcodcl(ifac,iv,1) = - fmul * vnrm * surfbo(2,ifac) / surfbn(ifac)
    rcodcl(ifac,iw,1) = - fmul * vnrm * surfbo(3,ifac) / surfbn(ifac)

    if (itytur.eq.2) then

      rcodcl(ifac,ik,1)  = cvar_k(iel)
      rcodcl(ifac,iep,1) = cvar_ep(iel)

    elseif (itytur.eq.3) then

        rcodcl(ifac,ir11,1) = cvar_rij(1,iel)
        rcodcl(ifac,ir22,1) = cvar_rij(2,iel)
        rcodcl(ifac,ir33,1) = cvar_rij(3,iel)
        rcodcl(ifac,ir12,1) = cvar_rij(4,iel)
        rcodcl(ifac,ir13,1) = cvar_rij(6,iel)
        rcodcl(ifac,ir23,1) = cvar_rij(5,iel)

      rcodcl(ifac,iep,1)  = cvar_ep(iel)

      if (iturb.eq.32) then
        rcodcl(ifac,ial,1)  = cvar_al(iel)
      endif

    elseif (itytur.eq.5) then

      rcodcl(ifac,ik,1)  = cvar_k(iel)
      rcodcl(ifac,iep,1) = cvar_ep(iel)
      rcodcl(ifac,iphi,1) = cvar_phi(iel)

      if (iturb.eq.50) then
        rcodcl(ifac,ifb,1)  = cvar_fb(iel)
      elseif (iturb.eq.51) then
        rcodcl(ifac,ial,1)  = cvar_al(iel)
      endif

    elseif (iturb.eq.60) then

      rcodcl(ifac,ik,1)  = cvar_k(iel)
      rcodcl(ifac,iomg,1) = cvar_omg(iel)

    elseif (iturb.eq.70) then

      rcodcl(ifac,inusa,1) = cvar_nusa(iel)

    endif

    ! Handle scalars (a correction similar to that of velocity is suggested
    !                 rather than the simpler code below)
    if (nscal.gt.0) then
      do ii = 1, nscal
        call field_get_val_s(ivarfl(isca(ii)), cvar_scal)
        rcodcl(ifac,isca(ii),1) = cvar_scal(iel)
      enddo
    endif

  enddo

endif
!< [example_1]

!--------
! Formats
!--------

!----
! End
!----

deallocate(lstelt)  ! temporary array for boundary faces selection

return
end subroutine cs_f_user_boundary_conditions

File: src/user_examples/cs_user_boundary_conditions-auto_inlet_profile.cpp
/*============================================================================
 * User definition of boundary conditions.
 *============================================================================*/

/* VERS */

/*
  This file is part of code_saturne, a general-purpose CFD tool.

  Copyright (C) 1998-2026 EDF S.A.

  This program is free software; you can redistribute it and/or modify it under
  the terms of the GNU General Public License as published by the Free Software
  Foundation; either version 2 of the License, or (at your option) any later
  version.

  This program is distributed in the hope that it will be useful, but WITHOUT
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
  FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
  details.

  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 51 Franklin
  Street, Fifth Floor, Boston, MA 02110-1301, USA.
*/

/*----------------------------------------------------------------------------*/

#include "cs_headers.h"

/*----------------------------------------------------------------------------
 * Standard library headers
 *----------------------------------------------------------------------------*/

#include <assert.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

/*----------------------------------------------------------------------------
 * Local headers
 *----------------------------------------------------------------------------*/

/*----------------------------------------------------------------------------*/
/*
 * \file cs_user_boundary_conditions-auto_inlet_profile.cpp
 *
 * \brief User functions for boundary condition definitions.
 */
/*----------------------------------------------------------------------------*/

/*=============================================================================
 * User function definitions
 *============================================================================*/

/*----------------------------------------------------------------------------*/
/*
 * User definition of boundary conditions.
 *
 * \param[in, out]  domain   pointer to a cs_domain_t structure
 * \param[in, out]  bc_type  boundary face types
 */
/*----------------------------------------------------------------------------*/

void
cs_user_boundary_conditions([[maybe_unused]] cs_domain_t  *domain,
                            [[maybe_unused]] int           bc_type[])
{
  /*! [init] */
  const cs_lnum_t n_cells_ext = domain->mesh->n_cells_with_ghosts;
  const cs_lnum_t n_b_faces = domain->mesh->n_b_faces;

  const int n_fields = cs_field_n_fields();
  const int keysca = cs_field_key_id("scalar_id");

  const int nt_cur = domain->time_step->nt_cur;

  const cs_lnum_t *b_face_cells = domain->mesh->b_face_cells;

  const cs_nreal_3_t *restrict b_face_u_normal
    = domain->mesh_quantities->b_face_u_normal;
  const cs_real_t *restrict b_face_surf
    = domain->mesh_quantities->b_face_surf;

  const cs_real_t viscl0 = cs_glob_fluid_properties->viscl0;

  const auto bpro_rho = CS_F_(rho_b)->get_val_s();
  const auto cvar_vel = CS_F_(vel)->get_val_v();

  const cs_zone_t  *zn = nullptr;
  /*! [init] */

  /* Assign a pseudo-periodic channel type inlet to a set of boundary faces.

   * For each subset:
   *   - use selection criteria to filter boundary faces of a given subset
   *   - loop on faces from a subset
   *   - set the boundary condition for each face
   *
   * A feedback loop is used so as to progressively reach a state similar
   * to that of a periodic channel at the inlet. */

  /*! [example_1] */
  zn = cs_boundary_zone_by_name("inlet");

  constexpr cs_real_t fmprsc = 1.; // mean prescribed velocity

  auto vel_val_ext = CS_F_(vel)->bc_coeffs->get_val_ext_v();

  if (nt_cur == 1) {

    /* For the Rij-EBRSM model (and possibly V2f), we need a non-flat profile,
     * so as to ensure turbulent production, and avoid laminarization;
     * here, we simply divide the initial velocity by 10 for inlet
     * faces adjacent to the wall.

     * The loop below assumes wall conditions have been defined first
     * (in the GUI, or in this file, before the current test). */

    cs_array<int> mrkcel;
    if ((cs_glob_turb_model->model == CS_TURB_RIJ_EPSILON_EBRSM) ||
        (cs_glob_turb_model->itytur == 5)){

      mrkcel.reshape(n_cells_ext);
      for (cs_lnum_t i = 0; i < n_cells_ext; i++)
        mrkcel[i] = 0;

      for (cs_lnum_t f_id = 0; f_id < n_b_faces; f_id++) {
        if (bc_type[f_id] == CS_SMOOTHWALL) {
          const cs_lnum_t c_id = b_face_cells[f_id];
          mrkcel[c_id] = 1;
        }
      }
    }

    for (cs_lnum_t e_idx = 0; e_idx < zn->n_elts; e_idx++) {

      const cs_lnum_t face_id = zn->elt_ids[e_idx];
      const cs_lnum_t c_id = b_face_cells[face_id];

      bc_type[face_id] = CS_INLET;

      for (int ii = 0; ii < 3; ii++)
        vel_val_ext(face_id, ii) = -fmprsc*b_face_u_normal[face_id][ii];

      if (mrkcel[c_id] == 1)
        vel_val_ext(face_id, 0) = fmprsc/10;

      cs_real_t uref2 = 0;
      for (int ii = 0; ii< CS_F_(vel)->dim; ii++)
        uref2 += cs_math_pow2(vel_val_ext(face_id, ii));
      uref2 = cs::max(uref2, 1e-12);

      /* Turbulence example computed using equations valid for a pipe.

       * We will be careful to specify a hydraulic diameter adapted
       *   to the current inlet.

       * We will also be careful if necessary to use a more precise
       *   formula for the dynamic viscosity use in the calculation of
       *   the Reynolds number (especially if it is variable, it may be
       *   useful to take the law from 'cs_user_physical_properties'
       *   Here, we use by default the 'viscl0" value.
       *   Regarding the density, we have access to its value at boundary
       *   faces (b_rho) so this value is the one used here (specifically,
       *   it is consistent with the processing in 'cs_user_physical_properties',
       *   in case of variable density) */

      /* Hydraulic diameter */
      cs_real_t xdh = 1.0;

      /* Calculation of turbulent inlet conditions using
         the turbulence intensity and standard laws for a circular pipe
         (their initialization is not needed here but is good practice) */

      cs_real_t b_rho = bpro_rho[face_id];

      cs_turbulence_bc_inlet_hyd_diam(face_id,
                                      uref2,
                                      xdh,
                                      b_rho,
                                      viscl0);
    }

    /* Initialize user scalars inlet */

    for (int f_id = 0; f_id < n_fields; f_id++) {
      cs_field_t *fld = cs_field(f_id);

      if (! (fld->type & (CS_FIELD_VARIABLE | CS_FIELD_USER)))
        continue;
      int sc_id = fld->get_key_int(keysca) - 1;
      if (sc_id < 0)
        continue;

      auto val_ext = fld->bc_coeffs->get_val_ext();

      for (cs_lnum_t e_idx = 0; e_idx < zn->n_elts; e_idx++) {
        const cs_lnum_t face_id = zn->elt_ids[e_idx];
        val_ext[face_id] = 1;
      }

    }
  }

  else {

    /* Subsequent time steps
     *----------------------*/

    cs_real_2_t acc = {0, 0};

    /* Accessors for BC values */

    cs_span<cs_real_t> k_val_ext;
    cs_span<cs_real_t> eps_val_ext;
    cs::mdspan<cs_real_t, 2, cs::layout::left> rij_val_ext;
    cs_span<cs_real_t> alp_bl_val_ext;
    cs_span<cs_real_t> phi_val_ext;
    cs_span<cs_real_t> f_bar_val_ext;
    cs_span<cs_real_t> omg_val_ext;
    cs_span<cs_real_t> nusa_val_ext;

    cs_span<cs_real_t> cvar_k;
    cs_span<cs_real_t> cvar_eps;
    cs_span_2d<cs_real_t> cvar_rij;
    cs_span<cs_real_t> cvar_alp_bl;
    cs_span<cs_real_t> cvar_phi;
    cs_span<cs_real_t> cvar_f_bar;
    cs_span<cs_real_t> cvar_omg;
    cs_span<cs_real_t> cvar_nusa;

    if (cs_glob_turb_model->itytur == 2) {
      k_val_ext = CS_F_(k)->bc_coeffs->get_val_ext();
      eps_val_ext = CS_F_(eps)->bc_coeffs->get_val_ext();
      cvar_k = CS_F_(k)->get_val_s();
      cvar_eps = CS_F_(eps)->get_val_s();
    }
    else if (cs_glob_turb_model->itytur == 3) {
      rij_val_ext = CS_F_(rij)->bc_coeffs->get_val_ext_t();
      eps_val_ext = CS_F_(eps)->bc_coeffs->get_val_ext();
      cvar_rij = CS_F_(rij)->get_val_t();
      cvar_eps = CS_F_(eps)->get_val_s();
      if (cs_glob_turb_model->model == CS_TURB_RIJ_EPSILON_EBRSM) {
        alp_bl_val_ext = CS_F_(alp_bl)->bc_coeffs->get_val_ext();
        cvar_alp_bl= CS_F_(alp_bl)->get_val_s();
      }
    }
    if (cs_glob_turb_model->itytur == 5) {
      k_val_ext = CS_F_(k)->bc_coeffs->get_val_ext();
      eps_val_ext = CS_F_(eps)->bc_coeffs->get_val_ext();
      phi_val_ext = CS_F_(phi)->bc_coeffs->get_val_ext();
      cvar_k = CS_F_(k)->get_val_s();
      cvar_eps = CS_F_(eps)->get_val_s();
      cvar_phi = CS_F_(phi)->get_val_s();
      if (cs_glob_turb_model->model == CS_TURB_V2F_PHI) {
        f_bar_val_ext = CS_F_(f_bar)->bc_coeffs->get_val_ext();
        cvar_f_bar = CS_F_(f_bar)->get_val_s();
      }
      else if (cs_glob_turb_model->model == CS_TURB_V2F_BL_V2K) {
        alp_bl_val_ext = CS_F_(alp_bl)->bc_coeffs->get_val_ext();
        cvar_alp_bl= CS_F_(alp_bl)->get_val_s();
      }
    }
    else if (cs_glob_turb_model->model == CS_TURB_K_OMEGA) {
      k_val_ext = CS_F_(k)->bc_coeffs->get_val_ext();
      omg_val_ext = CS_F_(omg)->bc_coeffs->get_val_ext();
      cvar_k = CS_F_(k)->get_val_s();
      cvar_omg = CS_F_(omg)->get_val_s();
    }
    else if (cs_glob_turb_model->model ==  CS_TURB_SPALART_ALLMARAS) {
      nusa_val_ext = CS_F_(nusa)->bc_coeffs->get_val_ext();
      cvar_nusa = CS_F_(nusa)->get_val_s();
    }

    /* Estimate multiplier */

    for (cs_lnum_t e_idx = 0; e_idx < zn->n_elts; e_idx++) {

      const cs_lnum_t face_id = zn->elt_ids[e_idx];
      const cs_lnum_t c_id = b_face_cells[face_id];

      const cs_real_t vnrm = cs_math_3_norm(cvar_vel.sub_array(c_id));

      acc[0] += vnrm*b_face_surf[face_id];
      acc[1] += b_face_surf[face_id];
    }

    cs_parall_sum(2, CS_REAL_TYPE, acc);

    cs_real_t fmul = 0; /* zero velocity in bulk domain */
    if (acc[0] > cs_math_epzero)
      fmul = fmprsc/(acc[0]/acc[1]); /* estimate flow multiplier */

    /* Apply BCs */
    for (cs_lnum_t e_idx = 0; e_idx < zn->n_elts; e_idx++) {
      const cs_lnum_t face_id = zn->elt_ids[e_idx];
      const cs_lnum_t c_id = b_face_cells[face_id];

      const cs_real_t vnrm = cs_math_3_norm(cvar_vel.sub_array(c_id));

      assert(bc_type[face_id] == CS_INLET);

      for (cs_lnum_t ii = 0; ii < 3; ii++)
        vel_val_ext(face_id, ii) = -fmul*vnrm*b_face_u_normal[face_id][ii];

      if (cs_glob_turb_model->itytur == 2) {
        k_val_ext[face_id] = cvar_k[c_id];
        eps_val_ext[face_id] = cvar_eps[c_id];
      }

      else if (cs_glob_turb_model->itytur == 3) {
        for (cs_lnum_t ii = 0; ii< 6; ii++)
          rij_val_ext(face_id, ii) = cvar_rij(c_id, ii);
        eps_val_ext[face_id] = cvar_eps[c_id];
        if (cs_glob_turb_model->model == CS_TURB_RIJ_EPSILON_EBRSM)
          alp_bl_val_ext[face_id] = cvar_alp_bl[c_id];
      }

      else if (cs_glob_turb_model->itytur == 5) {
        k_val_ext[face_id] = cvar_k[c_id];
        eps_val_ext[face_id] = cvar_eps[c_id];
        phi_val_ext[face_id] = cvar_phi[c_id];
        if (cs_glob_turb_model->model == CS_TURB_V2F_PHI)
          f_bar_val_ext[face_id] = cvar_f_bar[c_id];
        else if (cs_glob_turb_model->model == CS_TURB_V2F_BL_V2K)
          alp_bl_val_ext[face_id] = cvar_alp_bl[c_id];
      }

      else if (cs_glob_turb_model->model == CS_TURB_K_OMEGA) {
        k_val_ext[face_id] = cvar_k[c_id];
        omg_val_ext[face_id] = cvar_omg[c_id];
      }
      else if (cs_glob_turb_model->model ==  CS_TURB_SPALART_ALLMARAS) {
        nusa_val_ext[face_id] = cvar_nusa[c_id];
      }
    }

    for (int f_id = 0; f_id < n_fields; f_id++) {
      cs_field_t *fld = cs_field(f_id);

      if (! (fld->type & (CS_FIELD_VARIABLE | CS_FIELD_USER)))
        continue;
      int sc_id = fld->get_key_int(keysca) - 1;
      if (sc_id < 0)
        continue;

      auto val_ext = fld->bc_coeffs->get_val_ext();
      auto cvar = fld->get_val_s();

      for (cs_lnum_t e_idx = 0; e_idx < zn->n_elts; e_idx++) {
        const cs_lnum_t face_id = zn->elt_ids[e_idx];
        const cs_lnum_t c_id = b_face_cells[face_id];
        val_ext[face_id] = cvar[c_id];
      }
    }
  }
  /*! [example_1] */
}

/*----------------------------------------------------------------------------*/
